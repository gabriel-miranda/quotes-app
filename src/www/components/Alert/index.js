import React from 'react';

const ALERT_STYLES = {
  margin: '20px 0 0',
};

export default () => (
  <div className="alert alert-danger" style={ALERT_STYLES}>
    <p>Something went wrong. Please try again in a few moments.</p>
  </div>
);
