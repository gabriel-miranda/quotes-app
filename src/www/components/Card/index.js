import React from 'react';
import PropTypes from 'prop-types';

const CARD_STYLES = {
  margin: '20px 0',
  padding: '20px',
  borderRadius: '3px',
};

const Card = ({id, text, authorName}) => (
  <div className="card-docs" style={CARD_STYLES}>
    <p>Quote <strong>#{id}</strong></p>
    <p>"{text}"</p>
    <p>
      <strong>
        <em>
          - {authorName || 'Anonymous'}
        </em>
      </strong>
    </p>
  </div>
);

Card.propTypes = {
  id: PropTypes.number.isRequired,
  text: PropTypes.string.isRequired,
  authorName: PropTypes.string.isRequired,
};

export default Card;
