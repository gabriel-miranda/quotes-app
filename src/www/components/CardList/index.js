import React from 'react';
import PropTypes from 'prop-types';
import Waypoint from 'react-waypoint';
import Container from '../Layout/Container';
import Content from '../Layout/Content';
import Card from '../Card';
import ErrorMessage from '../ErrorMessage';
import NoResults from '../NoResults';
import Spinner from '../Spinner';

const isLoading = data => (
  !data || !data.pagination || data.pagination.page !== data.pagination.pageCount
);

const haveNoResults = data => (
  !isLoading(data) && !data.results.length
);

const CardItem = props => <Card key={props.id} {...props} />;

const CardList = ({
  data,
  error,
  fetching,
  loadNextPage,
}) => (
  <Container>
    <Content>
      {data &&
        data.results.map(CardItem)
      }
      {haveNoResults(data) &&
        <NoResults />
      }
      {error &&
        <ErrorMessage />
      }
      {isLoading(data) &&
        <Spinner />
      }
      {(!fetching && isLoading(data)) &&
        <Waypoint
          key="quotes"
          onEnter={loadNextPage}
        />
      }
    </Content>
  </Container>
);

CardItem.propTypes = {
  id: PropTypes.string.isRequired,
};

CardList.propTypes = {
  data: PropTypes.object,
  error: PropTypes.object,
  fetching: PropTypes.bool,
  loadNextPage: PropTypes.func.isRequired,
};

CardList.defaultProps = {
  data: null,
  error: null,
  fetching: true,
};

export default CardList;
