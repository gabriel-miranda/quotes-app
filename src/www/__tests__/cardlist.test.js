/* eslint-env jest */

import { shallow } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

import CardList from '../components/CardList';
import Card from '../components/Card';
import Alert from '../components/Alert';

const data = {
  results: [
    {id: 1, text: 'Test', authorName: ''},
    {id: 2, text: 'Test', authorName: ''},
  ],
};

const error = new Error('Test error');

describe('CardList test with enzyme', () => {
  it('should show an error if an error happened', () => {
    const cardList = shallow(<CardList />);
    cardList.setProps({data, error});
    expect(cardList.contains(<Alert />)).toEqual(true);
  });

  it('should not show an error if no error is provided', () => {
    const cardList = shallow(<CardList />);
    cardList.setProps({data});
    expect(cardList.contains(<Alert />)).toEqual(false);
  });

  it('should show a card if data is provided', () => {
    const cardList = shallow(<CardList />);
    cardList.setProps({data});
    expect(cardList.find(Card).length).toBeGreaterThan(0);
  });
});

describe('CardList test with Snapshot Testing', () => {
  it('should show cardlist correctly', () => {
    const component = renderer.create(<CardList />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should show cardlist with props correctly', () => {
    const component = renderer.create(<CardList data={data} error={error} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
