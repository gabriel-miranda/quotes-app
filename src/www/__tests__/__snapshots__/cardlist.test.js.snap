// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardList test with Snapshot Testing should show cardlist correctly 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 col-md-offset-3"
    >
      <div
        className="spinner spinner-lg is-auth0"
        style={
          Object {
            "margin": "0 auto 20px",
          }
        }
      >
        <div
          className="circle"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`CardList test with Snapshot Testing should show cardlist with props correctly 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 col-md-offset-3"
    >
      <div
        className="card-docs"
        style={
          Object {
            "borderRadius": "3px",
            "margin": "20px 0",
            "padding": "20px",
          }
        }
      >
        <p>
          Quote 
          <strong>
            #
            91
          </strong>
        </p>
        <p>
          "
          It takes both sunshine and rain to make a rainbow.
          "
        </p>
        <p>
          <strong>
            -
            <em>
              Anonymous
            </em>
          </strong>
        </p>
      </div>
      <div
        className="card-docs"
        style={
          Object {
            "borderRadius": "3px",
            "margin": "20px 0",
            "padding": "20px",
          }
        }
      >
        <p>
          Quote 
          <strong>
            #
            92
          </strong>
        </p>
        <p>
          "
          A beautiful thing is never perfect.
          "
        </p>
        <p>
          <strong>
            -
            <em>
              Anonymous
            </em>
          </strong>
        </p>
      </div>
      <div
        className="card-docs"
        style={
          Object {
            "borderRadius": "3px",
            "margin": "20px 0",
            "padding": "20px",
          }
        }
      >
        <p>
          Quote 
          <strong>
            #
            37
          </strong>
        </p>
        <p>
          "
          From small beginnings come great things.
          "
        </p>
        <p>
          <strong>
            -
            <em>
              Anonymous
            </em>
          </strong>
        </p>
      </div>
      <div
        className="card-docs"
        style={
          Object {
            "borderRadius": "3px",
            "margin": "20px 0",
            "padding": "20px",
          }
        }
      >
        <p>
          Quote 
          <strong>
            #
            90
          </strong>
        </p>
        <p>
          "
          The day is already blessed, find peace within it.
          "
        </p>
        <p>
          <strong>
            -
            <em>
              Anonymous
            </em>
          </strong>
        </p>
      </div>
      <div
        className="card-docs"
        style={
          Object {
            "borderRadius": "3px",
            "margin": "20px 0",
            "padding": "20px",
          }
        }
      >
        <p>
          Quote 
          <strong>
            #
            16
          </strong>
        </p>
        <p>
          "
          Today is the tomorrow we worried about yesterday.
          "
        </p>
        <p>
          <strong>
            -
            <em>
              Anonymous
            </em>
          </strong>
        </p>
      </div>
      <div
        className="card-docs"
        style={
          Object {
            "borderRadius": "3px",
            "margin": "20px 0",
            "padding": "20px",
          }
        }
      >
        <p>
          Quote 
          <strong>
            #
            18
          </strong>
        </p>
        <p>
          "
          Every man dies. Not every man really lives.
          "
        </p>
        <p>
          <strong>
            -
            <em>
              Anonymous
            </em>
          </strong>
        </p>
      </div>
      <div
        className="card-docs"
        style={
          Object {
            "borderRadius": "3px",
            "margin": "20px 0",
            "padding": "20px",
          }
        }
      >
        <p>
          Quote 
          <strong>
            #
            28
          </strong>
        </p>
        <p>
          "
          The day is already blessed, find peace within it.
          "
        </p>
        <p>
          <strong>
            -
            <em>
              Anonymous
            </em>
          </strong>
        </p>
      </div>
      <div
        className="card-docs"
        style={
          Object {
            "borderRadius": "3px",
            "margin": "20px 0",
            "padding": "20px",
          }
        }
      >
        <p>
          Quote 
          <strong>
            #
            17
          </strong>
        </p>
        <p>
          "
          It's easier to see the mistakes on someone else's paper.
          "
        </p>
        <p>
          <strong>
            -
            <em>
              Anonymous
            </em>
          </strong>
        </p>
      </div>
      <div
        className="card-docs"
        style={
          Object {
            "borderRadius": "3px",
            "margin": "20px 0",
            "padding": "20px",
          }
        }
      >
        <p>
          Quote 
          <strong>
            #
            113
          </strong>
        </p>
        <p>
          "
          Some pursue happiness, others create it.
          "
        </p>
        <p>
          <strong>
            -
            <em>
              Anonymous
            </em>
          </strong>
        </p>
      </div>
      <div
        className="card-docs"
        style={
          Object {
            "borderRadius": "3px",
            "margin": "20px 0",
            "padding": "20px",
          }
        }
      >
        <p>
          Quote 
          <strong>
            #
            107
          </strong>
        </p>
        <p>
          "
          Every new day is another chance to change your life.
          "
        </p>
        <p>
          <strong>
            -
            <em>
              Anonymous
            </em>
          </strong>
        </p>
      </div>
      <div
        className="alert alert-danger"
        style={
          Object {
            "margin": "20px 0 0",
          }
        }
      >
        <p>
          Something went wrong. Please try again in a few moments.
        </p>
      </div>
      <div
        className="spinner spinner-lg is-auth0"
        style={
          Object {
            "margin": "0 auto 20px",
          }
        }
      >
        <div
          className="circle"
        />
      </div>
      <span
        style={
          Object {
            "fontSize": 0,
          }
        }
      />
    </div>
  </div>
</div>
`;
